src/libstore/build/derivation-goal.cc

```diff
@@ -1019,7 +1020,45 @@ void DerivationGoal::buildDone()
 }
 
 void DerivationGoal::resolvedFinished() {
-    done(BuildResult::Built);
+    assert(resolvedDrv);
+
+    // If the derivation was originally a full `Derivation` (and not just
+    // a `BasicDerivation`, we must retrieve it because the `staticOutputHashes`
+    // will be wrong otherwise
+    Derivation fullDrv = *drv;
+    if (auto upcasted = dynamic_cast<Derivation *>(drv.get()))
+        fullDrv = *upcasted;
```

Comment:
This seems a bit ugly... It makes it hard to reason about `DerivationGoal` if it sneakily does something different if the `drv` (which is a `BasicDerivation`) is not *really* a `BasicDerivation`. E.g. what does this mean for the remote build case (which is the case where we receive a `BasicDerivation`)? 

Maybe it's cleaner to handle this in the `DerivationGoal` constructors and `getDerivation` method. I.e. add a field `originalHashes` to `DerivationGoal`, which gets initialized in `DerivationGoal::DerivationGoal(... BasicDerivation ...)` from the `BasicDerivation`, and in `DerivationGoal::getDerivation` from the full derivation.

Reply:
Fixed in 3c8b629c03f4e55835508119b8c4fca6bb428ef9